### PR TITLE
Prepare 16.0.0

### PR DESCRIPTION
**Pull request by mattxwang**
Ref #6930

~~(draft PR for breaking v16 changes)~~

### CODE REVIEW

**Comment by ybiquitous** [comment]
Hi @mattxwang, thanks for creating this PR so we can easily track v16 changes. 👍🏼 

Just to confirm, we will need to regularly rebase the `main` branch, right?

**Comment by mattxwang** [comment]
Sorry all, I've been a bit MIA as work has been taking up most of my time. Noticed that there's a minor coverage drop; is that something we should be concerned about (I could take a look at adding a missing test)

**Comment by ybiquitous** [comment]
I think there is such a minor coverage drop because `.mjs` and `.cjs` files are mixed now. This will be resolved when we finish the migration to ESM. But surely we need to add test cases for CJS.

**Comment by ybiquitous** [comment]
Finally, #7160 should be the last PR to merge into `v16`. After merging #7160, we will be ready to merge this PR.

Although we may need a few changes in the documentation before the 16.0.0 release, we can also do that on the `main` branch.

**Review by ybiquitous — APPROVED** [review]
Now, we should be ready to merge. LGTM.
I want one more approval, just in case, since this is a big PR. 🤯 

**Note to self: we must NOT "squash and merge", but just "merge"!**

File: types/stylelint/index.d.ts
Comment (Mouvedia) [inline types/stylelint/index.d.ts:605]: minor
non-blocking

This union should have been generated from a `Set` imported from a file in the reference directory.

Comment (ybiquitous) [reply types/stylelint/index.d.ts:605]: Yeah, syncing the implementation and type definition is best. I don't know how to achieve it, though. 😅

Comment (Mouvedia) [reply types/stylelint/index.d.ts:605]: It should look something like that:

```
export const bar = /** @type {const} */ (['animation', 'background', 'etc']);
…
import { bar } from 'qux.mjs';
type Baz = typeof foo[number]; // 'animation' | 'background' | 'etc'
```

Comment (ybiquitous) [reply types/stylelint/index.d.ts:605]: Let's discuss this in another place.

**Review by jeddy3 — APPROVED** [review]
Amazing work @ybiquitous! This was a colossal effort.

Let's merge so that we can release `next`, start testing, and patch any bugs on `main` before releasing `16.0.0`.

**Pull request merged**
